linera-base/src/crypto/mod.rs

```diff
@@ -197,42 +233,20 @@ impl AccountPublicKey {
 
 impl AccountSignature {
     /// Verifies the signature for the `value` using the provided `public_key`.
-    pub fn verify<'de, T>(&self, value: &T, author: AccountPublicKey) -> Result<(), CryptoError>
+    pub fn verify<'de, T>(&self, value: &T) -> Result<(), CryptoError>
     where
         T: BcsSignable<'de> + std::fmt::Debug,
     {
-        match (self, author) {
-            (AccountSignature::Ed25519(signature), AccountPublicKey::Ed25519(public_key)) => {
-                signature.check(value, public_key)
-            }
-            (AccountSignature::Secp256k1(signature), AccountPublicKey::Secp256k1(public_key)) => {
-                signature.check(value, &public_key)
-            }
-            (
-                AccountSignature::EvmSecp256k1(signature),
-                AccountPublicKey::EvmSecp256k1(public_key),
-            ) => signature.check(value, &public_key),
-            (AccountSignature::Ed25519(_), _) => {
-                let type_name = std::any::type_name::<T>();
-                Err(CryptoError::InvalidSignature {
-                    error: "invalid signature scheme. Expected Ed25519 signature.".to_string(),
-                    type_name: type_name.to_string(),
-                })
-            }
-            (AccountSignature::Secp256k1(_), _) => {
-                let type_name = std::any::type_name::<T>();
-                Err(CryptoError::InvalidSignature {
-                    error: "invalid signature scheme. Expected secp256k1 signature.".to_string(),
-                    type_name: type_name.to_string(),
-                })
-            }
-            (AccountSignature::EvmSecp256k1(_), _) => {
-                let type_name = std::any::type_name::<T>();
-                Err(CryptoError::InvalidSignature {
-                    error: "invalid signature scheme. Expected EvmSecp256k1 signature.".to_string(),
-                    type_name: type_name.to_string(),
-                })
-            }
+        match self {
+            AccountSignature::Ed25519 {
+                signature,
+                public_key,
+            } => signature.check(value, public_key),
+            AccountSignature::Secp256k1 {
+                signature,
+                public_key,
+            } => signature.check(value, public_key),
+            AccountSignature::EvmSecp256k1 { signature, .. } => signature.check_with_recover(value),
```

Comment:
The fact that we're not using the address is suspicious

Reply:
I agree, I was on the fence about this but if others think so as well - I'll do it. It should address the other comment as well.

Reply:
Done in [Verify that recovered address matches the sender](https://github.com/linera-io/linera-protocol/pull/4088/commits/4de775be41e8c589716e6c6ac0b7742e38ef9336)